Lib/asyncio/pools.py

```diff
@@ -0,0 +1,173 @@
+
+    asyncio.run(main())
+    """
+
+    def __init__(self, concurrency=None):
+        if concurrency is None:
+            concurrency = min(32, (os.cpu_count() or 1) + 4)
+
+        self._concurrency = concurrency
+        self._running = False
+        self._closed = False
+        self._loop = None
+        self._pool = None
+
+    async def astart(self):
+        self._loop = events.get_running_loop()
+        await self._spawn_threadpool()
+
+    async def __aenter__(self):
+        await self.astart()
+        return self
+
+    async def aclose(self):
+        await self._shutdown_threadpool()
+
+    async def __aexit__(self, exc_type, exc_value, exc_traceback):
+        await self.aclose()
+
+    async def run(self, /, func, *args, **kwargs):
+        if not self._running:
+            raise RuntimeError(f"unable to run {func!r}, "
+                               "threadpool is not running")
+
+        func_call = functools.partial(func, *args, **kwargs)
+        executor = self._pool
+        return await futures.wrap_future(
+            executor.submit(func_call), loop=self._loop)
+
+    async def _spawn_threadpool(self):
+        """Schedule the spawning of the threadpool.
+
+        Asynchronously spawns a threadpool with *concurrency* threads.
+        """
+        if self._running:
+            raise RuntimeError("threadpool is already running")
+
+        if self._closed:
+            raise RuntimeError("threadpool is closed")
+
+        future = self._loop.create_future()
+        thread = threading.Thread(target=self._do_spawn, args=(future,))
+        thread.start()
+        try:
+            await future
+        finally:
+            thread.join()
+
+    def _do_spawn(self, future):
+        try:
+            self._pool = concurrent.futures.ThreadPoolExecutor(
+                                                max_workers=self._concurrency)
+            self._running = True
+            self._loop.call_soon_threadsafe(future.set_result, None)
+        except Exception as ex:
+            self._loop.call_soon_threadsafe(future.exception, ex)
+
+    async def _shutdown_threadpool(self):
+        """Schedule the shutdown of the threadpool.
+
+        Asynchronously joins all of the threads in the threadpool.
+        """
+        if self._closed:
+            raise RuntimeError("threadpool is already closed")
+
+        # Set _running to False as early as possible
+        self._running = False
+
+        future = self._loop.create_future()
+        thread = threading.Thread(target=self._do_shutdown, args=(future,))
+        thread.start()
+        try:
+            await future
+        finally:
+            thread.join()
+
+    def _do_shutdown(self, future):
+        try:
+            self._pool.shutdown()
+            self._closed = True
+            self._loop.call_soon_threadsafe(future.set_result, None)
+        except Exception as ex:
+            self._loop.call_soon_threadsafe(future.exception, ex)
```

Comment:
```suggestion
            self._loop.call_soon_threadsafe(future.set_exception, ex)
```

Reply:
Would be cool to figure out if we can test this.

Reply:
Thanks, good catch. I don't frequently use the low-level API for `asyncio.Future` in my own code.

@1st1 
> Would be cool to figure out if we can test this.

Do you have something specific in mind, such as ensuring that exceptions are correctly propagated to the result of `await run()`? That should be doable and fairly straightforward. It would make sense IMO to add a test for that.

Otherwise, it might be a bit difficult to directly ensure the exception was correctly set to the future without creating a separate mocked version to use for the tests, since the future is internal to its respective method.